cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt

```diff
@@ -774,24 +765,35 @@ class ScopeManager : ScopeProvider {
         searchScope: Scope?,
         stopIfFound: Boolean = false,
         noinline predicate: (T) -> Boolean
+    ): List<T> {
+        return resolve(T::class.java, searchScope, stopIfFound, predicate)
+    }
+
+    fun <T : Declaration> resolve(
+        klass: Class<T>,
+        searchScope: Scope?,
+        stopIfFound: Boolean = false,
+        predicate: (T) -> Boolean
     ): List<T> {
         var scope = searchScope
         val declarations = mutableListOf<T>()
 
         while (scope != null) {
             if (scope is ValueDeclarationScope) {
-                declarations.addAll(scope.valueDeclarations.filterIsInstance<T>().filter(predicate))
+                declarations.addAll(
+                    scope.valueDeclarations.filterIsInstance(klass).filter(predicate)
+                )
             }
 
             if (scope is StructureDeclarationScope) {
-                var list = scope.structureDeclarations.filterIsInstance<T>().filter(predicate)
+                var list = scope.structureDeclarations.filterIsInstance(klass).filter(predicate)
 
                 // this was taken over from the old resolveStructureDeclaration.
                 // TODO(oxisto): why is this only when the list is empty?
```

Comment:
We can't use `list.addAll` because what we return from `filter(predicate)` is a `List`, not a `MutableList`, in any case it would not make a difference because `list` is empty at line 796 line. 

BUT in both cases the contents of `list` are already added to the `declarations` list using `addAll` and only `declarations` is returned, so it is not only returning the last matched template.

Reply:
Ah ok I missed the `declarations.addAll(list)`. Can we update the Todo comment then? Or is there still information missing on why this is done when the list is empty

Reply:
It is basically still copy paste from very old legacy code and I have no idea why it is done this way. Therefore I would keep this here and hopefully somebody somewhere will overhaul the template code.